text/3838-repr-scalable.md

```diff
@@ -0,0 +1,386 @@
+
+```rust
+fn sve_add(in_a: Vec<f32>, in_b: Vec<f32>, out_c: &mut Vec<f32>) {
+    let len = in_a.len();
+    unsafe {
+        // `svcntw` returns the actual number of elements that are in a 32-bit
+        // element vector
+        let step = svcntw() as usize;
+        for i in (0..len).step_by(step) {
+            let a = in_a.as_ptr().add(i);
+            let b = in_b.as_ptr().add(i);
+            let c = out_c as *mut f32;
+            let c = c.add(i);
+
+            // `svwhilelt_b32` generates a mask based on comparing the current
+            // index against the `len`
+            let pred = svwhilelt_b32(i as _, len as _);
+
+            // `svld1_f32` loads a vector register with the data from address
+            // `a`, zeroing any elements in the vector that are masked out
+            let sva = svld1_f32(pred, a);
+            let svb = svld1_f32(pred, b);
+
+            // `svadd_f32_m` adds `a` and `b`, any lanes that are masked out will
+            // take the keep value of `a`
+            let svc = svadd_f32_m(pred, sva, svb);
+
+            // `svst1_f32` will store the result without accessing any memory
+            // locations that are masked out
+            svst1_f32(svc, pred, c);
+        }
+    }
+}
+```
+
+From a user's perspective, writing code for scalable vectors isn't too different
+from when writing code with a fixed sized vector.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Types annotated with the `#[repr(simd)]` attribute contains either an array
+field or multiple fields to indicate the intended size of the SIMD vector that
+the type represents.
+
+Similarly, a `scalable(N)` representation is introduced to define a scalable
+vector type. `scalable(N)` accepts an integer to determine the minimum number of
+elements the vector contains. For example:
+
+```rust
+#[repr(simd, scalable(4))]
+pub struct svfloat32_t { _ty: [f32], }
+```
+
+As with the existing `repr(simd)`, `_ty` is purely a type marker, used to get
+the element type for the codegen backend.
+
+`svfloat32_t` is a scalable vector with a minimum of four `f32` elements and
+potentially more depending on the length of the vector register at runtime.
+
+## Choosing `N`
+[choosing-n]: #choosing-n
+
+Many intrinsics using scalable vectors accept both a predicate vector argument
+and data vector arguments. Predicate vectors determine whether a lane is on or
+off for the operation performed by any given intrinsic. Predicate vectors may
+use different registers of sizes to the vectors containing data.
+`repr(scalable)` is used to define vectors containing both data and predicates.
+
+As `repr(scalable(N))` is intended to be a permanently unstable attribute, any
+value of `N` is accepted by the attribute and it is the responsibility of
+whomever is defining the type to provide a valid value. A correct value for `N`
+depends on the purpose of the specific scalable vector type and the
+architecture.
```

Comment:
This kind of abdication of responsibility in practice, when the rules are simply not that hard to work out, just leaves it harder to maintain the compiler in the future as it amounts to saying "We don't have to write down any reasoning about validity, not even in documentation, and we will allow someone in the future to discover why our decisions were fragile and sometimes incorrect without protecting them against more easily-induced miscompilations". I realize rust-lang has accepted this kind of reasoning in the past, but it sucks ass and amounts to pissing in the eyes of future maintainers once you can no longer ask the previous ones due to their inevitable burnout.

Reply:
I'd prefer if you engaged on this RFC without describing decisions you disagree with as  "sucks ass" or "pissing in the eyes of future maintainers". I'm open to discussing the details of this proposal and changing them, as with any proposal, but I'm much more open to that when concerns are phrased [like this concern was in Ralf's comment](https://github.com/rust-lang/rfcs/pull/3838#discussion_r2205305466) rather than how you've chosen to phrase yours.